stdlib/public/core/StringWordBreaking.swift

```diff
@@ -0,0 +1,690 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension _StringGuts {
+  @inline(__always)
+  internal func roundDownToNearestWord(
+    _ i: String._WordView.Index
+  ) -> String._WordView.Index {
+    _internalInvariant(i._encodedOffset <= count)
+
+    let offset = i._encodedOffset
+
+    if offset == 0 || offset == count {
+      return i
+    }
+
+    return _slowRoundDownToNearestWord(i)
+  }
+
+  @inline(never)
+  internal func _slowRoundDownToNearestWord(
+    _ i: String._WordView.Index
+  ) -> String._WordView.Index {
+    let offset = i._encodedOffset
+    let start = _opaquePreviousWordIndex(endingAt: offset)
+    let stride = _opaqueNextWordIndex(startingAt: start) &- start
+    _internalInvariant(offset <= start + stride, "Word breaking inconsistency")
+
+    if offset >= start + stride {
+      return i
+    }
+
+    return String._WordView.Index(_encodedOffset: start)
+  }
+
+  @inline(never)
+  @_effects(releasenone)
+  internal func _opaqueNextWordIndex(startingAt i: Int) -> Int {
+    if _slowPath(isForeign) {
+      return _foreignOpaqueNextWordIndex(startingAt: i)
+    }
+
+    return withFastUTF8 { utf8 in
+      nextWordBoundary(startingAt: i) {
+        _internalInvariant($0 >= 0)
+
+        guard $0 < utf8.count else {
+          return nil
+        }
+
+        let (scalar, len) = _decodeScalar(utf8, startingAt: $0)
+        return (scalar, $0 &+ len)
+      }
+    }
+  }
+
+  internal func _foreignOpaqueNextWordIndex(startingAt i: Int) -> Int {
+#if _runtime(_ObjC)
+    return nextWordBoundary(startingAt: i) {
+      _internalInvariant($0 >= 0)
+
+      guard $0 < count else {
+        return nil
+      }
+
+      let scalars = String.UnicodeScalarView(self)
+      let idx = String.Index(_encodedOffset: $0)
+
+      let scalar = scalars[idx]
+      let nextIndex = scalars.index(after: idx)
+
+      return (scalar, nextIndex._encodedOffset)
+    }
+#else
+    fatalError("No foreign strings on this platform in this version of Swift.")
+#endif
+  }
+
+  internal func _opaquePreviousWordIndex(endingAt i: Int) -> Int {
+    if _slowPath(isForeign) {
+      return _foreignOpaquePreviousWordIndex(endingAt: i)
+    }
+
+    return withFastUTF8 { utf8 in
+      previousWordBoundary(endingAt: i) {
+        _internalInvariant($0 <= count)
+
+        guard $0 > 0 else {
+          return nil
+        }
```

Comment:
Is every call site expected to put these checks in? Why not have that inside the function itself then?

Reply:
I fully expect we'll need to have separate entry point for slices, where we compare against the slice's start index.

Reply:
Unless we make it public as `some Sequence<Substring>` which avoids slicing